### PR TITLE
test: Refactor subtree exclusion in lint tests

### LINKED ISSUE #17413: Subtree exclude mess in linters and update scripts
For various operations, it is important to exclude subtrees that have imported code:

- Doxygen: prevent cluttering our documentation with documentation that belongs with that library itself
- Scripted-diff: commits are not allowed to touch subtrees, this is easy to accidentally do with scripted changes
- Linters: some linters check policies that do not apply to upstream projects
- Test coverage: distorts the results; low test coverage of subsidiary libraries does not mean anything

It would be good for consistency, and to reduce developer frustration, for this list to be in one place only, then sourced from where it is needed.

For example, for the `crc32c` subtree I had to update these places:

- `contrib/devtools/copyright_header.py` (`EXCLUDE_DIRS`)
- `doc/Doxyfile.in` (`EXCLUDE`)
- `test/lint/lint-include-guards.py` (`EXCLUDE_FILES_WITH_PREFIX`)
- `test/lint/lint-includes.py` (`EXCLUDED_DIRS`)
- `test/lint/lint-python-utf8-encoding.py` (`EXCLUDED_DIRS`)
- `test/lint/lint-spelling.py` (`FILES_ARGS`, git ls-files inline)
- `test/lint/lint-whitespace.py` (`EXCLUDED_DIRS`)
- `Makefile.am` (`LCOV_FILTER_PATTERN`)

I'm not yet sure whether I forgot any places (because the excludes are specified in different formats in different places).

### Good first issue

The purpose of the `good first issue` label is to highlight which issues are suitable for a new contributor without a deep understanding of the codebase.

Useful skills: *shell scripting*, *python*

Want to work on this issue?

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by BrandonOdiwuor**
Fixes https://github.com/bitcoin/bitcoin/issues/17413

Refactor subtree exclusion in lint tests to one place

Second attempt after PR: https://github.com/bitcoin/bitcoin/pull/24435

### CODE REVIEW

File: test/lint/lint-python-utf8-encoding.py [hunk 1]
```diff
@@ -12,11 +12,10 @@
 
 from subprocess import check_output, CalledProcessError
 
-EXCLUDED_DIRS = ["src/crc32c/", "src/secp256k1/"]
```
Comment (fjahr) [inline test/lint/lint-python-utf8-encoding.py:15]: This is not the same list as what's in the file so you are now excluding more than before

**Review by fjahr — COMMENTED** [review]
This doesn't feel like much of an improvement as is. We are trading a tiny bit of deduplication for the txt parsing code that is much harder to reason about than the straightforward lists. Maybe there is a simpler way of doing this through a shared python module or so.

This would also need documentation so that people know where to add which dirs for exclusion if there are changes in the future.

File: test/lint/lint-python-utf8-encoding.py (hunk 1 above)
Comment (BrandonOdiwuor) [reply test/lint/lint-python-utf8-encoding.py]: fixed

**Comment by BrandonOdiwuor** [comment]
@fjahr I have changed this to a python script as suggested which is relatively easy to import and is more straight forward

- also added a doc

File: test/lint/lint_ignore_dirs.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+EXCLUDED_DIRS = ["src/leveldb/",
```
Comment (fjahr) [inline test/lint/lint_ignore_dirs.py:1]: Please name it `SHARED_EXCLUDED_DIRS`, that is clearer and I don't like that the same var name is reused in the whitespace and includes linters (also means it's harder to grep for).

**Review by fjahr — COMMENTED** [review]
Approach ACK, thanks, I like this a lot more.

File: test/lint/lint_ignore_dirs.py (hunk 2 above)
Comment (BrandonOdiwuor) [reply test/lint/lint_ignore_dirs.py]: renamed `EXCLUDED_DIRS` to `SHARED_EXCLUDED_DIRS`

**Comment by fjahr** [comment]
utACK 6e80169227ec7c8e624308c5e97a68760e10c231

**Comment by hebasto** [comment]
Concept ACK.

**Review by maflcko — COMMENTED** [review]
There is also `fn get_pathspecs_exclude_subtrees() -> Vec<String>`. The changes here seem fine. However, an alternative might be to rewrite the 3 lint checks to rust, so that they can use the already existing function. No strong opinion, though.

File: test/lint/lint_ignore_dirs.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+SHARED_EXCLUDED_DIRS = ["src/leveldb/",
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/lint/lint_ignore_dirs.py]: updated the var to `SHARED_EXCLUDED_SUBTREES`

**Comment by davidgumberg** [comment]
ACK https://github.com/bitcoin/bitcoin/commit/80fa7da21c470302165c47cc4a6a62fb44f997ef

I've written a draft PR to explore rewriting these tests in Rust: #29744

**Comment by fjahr** [comment]
> Rust is type-safe, so many issues that are found after code is written, reviewed, and merged just can not happen and will be caught, ideally before a pull is reviewed and merged. For example [#29068 (comment)](https://github.com/bitcoin/bitcoin/pull/29068#issuecomment-1852746032) would have failed early in Rust (`error[E0308]: mismatched types: expected i32, found bool`). Let me know if I should provide more examples that were hit in this project. I know that python has type annotations, but if they are not used (like in the past where type-bugs have bitten this project), they are useless. In Rust, it is impossible to forget to annotate the type.

I am aware of the general differences between Rust and Python. If there is consensus that rewriting Python programs to Rust is considered an improvement in general just because of the different language, then we should start with the functional test framework because that sees a lot more usage in terms of runs and code changes, no? 

But I was rather asking about other arguments that are specific to the linters as was discussed, like the reuse of the function mentioned already.

**Pull request merged**